### PR TITLE
add_ppminilm_demo

### PR DESCRIPTION
**Pull request by RachelXu7**
demo of PPMiniLM

### CODE REVIEW

File: demo/auto-compression/demo_ppminilm.py [hunk 1]
```diff
@@ -0,0 +1,239 @@
+import os
```
Comment (qingqing01) [inline demo/auto-compression/demo_ppminilm.py]: 统一放到 demo/auto-compression/pp-minilm 文件夹下

File: demo/auto-compression/pp-minilm/README.md [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+
+
+### 多策略融合压缩
+
+#### 1. 进行剪枝蒸馏压缩
+#### 1.1 配置config
+
+使用自动压缩进行剪枝蒸馏训练需要准备config文件，包括蒸馏、剪枝和训练三部分参数。以下示例以AFQMC数据集为例介绍，configs文件夹下可查看其他任务的配置文件。
+
+- 蒸馏参数
```
Comment (ceci3) [inline demo/auto-compression/pp-minilm/README.md]: image
这些位置的换行不对

File: demo/auto-compression/pp-minilm/README.md [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+蒸馏参数包括teacher网络模型路径（即微调后未剪枝的模型），自动压缩策略会自动查找教师网络节点和对应的学生网络节点进行蒸馏，不需要手动设置。
+
+```yaml
+Distillation:
+  teacher_model_dir: ./ppminilm-6l-768h/models/AFQMC
+  teacher_model_filename: inference.pdmodel
+  teacher_params_filename: inference.pdiparams
+```
+
+- 剪枝参数
```
Comment (ceci3) [inline demo/auto-compression/pp-minilm/README.md]: 同上

File: demo/auto-compression/pp-minilm/README.md [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+剪枝参数包括裁剪算法和裁剪度。
+```yaml
+Prune:
+  prune_algo: transformer_pruner
+  pruned_ratio: 0.25
+```
+
+- 训练参数
```
Comment (ceci3) [inline demo/auto-compression/pp-minilm/README.md]: 同上

File: demo/auto-compression/pp-minilm/README.md [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+  max_quant_count: 20
+  ptq_algo:
+  - KL
+  - hist
+  weight_quantize_type:
+  - channel_wise_abs_max
+```
+- 量化参数
+
+量化参数主要设置量化比特数和量化op类型，其中量化op包含卷积层（conv2d, depthwise_conv2d）和全连接层（matmul，matmul_v2）。
```
Comment (ceci3) [inline demo/auto-compression/pp-minilm/README.md]: 全连接层这里写的是 matmul，下面写的是mul

File: demo/auto-compression/pp-minilm/README.md [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+| 压缩前 | 74.03 | 56.66 | 60.21 | 80.98 | 76.20 | 84.21 | 77.36 | 72.81 |
+| 压缩后 | 73.56 | 56.38 | 59.87 | 80.80 | 76.44 | 82.23 | 77.77 | 72.44 |
+
+
+模型在不同任务上平均精度以及加速对比如下：
+|  PP-MiniLM | Accuracy（avg） | 时延(ms) | 加速比 |
+|:-------:|:----------:|:------------:| :------:|
+| 压缩前 |  72.81 | 128.01（FP32）27.9（FP16） | - |
+| 压缩后 |  72.44 | 17.97 | 612% |
+
```
Comment (ceci3) [inline demo/auto-compression/pp-minilm/README.md]: 这两个表格整合到一个里面？感觉这个表格出现好几次

Comment (RachelXu7) [reply demo/auto-compression/pp-minilm/README.md]: 开头结尾repeat了一次。那放开头还是结尾还是结尾呢

File: demo/auto-compression/pp-minilm/configs/afqmc_cfg.yaml [hunk 7]
```diff
@@ -0,0 +1,8 @@
+TrainConfig:
```
Comment (qingqing01) [inline demo/auto-compression/pp-minilm/configs/afqmc_cfg.yaml]: 都在 configs下面， _cfg 后缀可以去掉

**Review by yghstill — COMMENTED** [review]
auto-compression重命名为auto_compression

File: demo/auto-compression/pp-minilm/README.md [hunk 8]
```diff
@@ -0,0 +1,128 @@
+# PP-MiniLM模型自动压缩示例
+
+本示例介绍了对PaddleNLP中PP-MiniLM模型进行自动压缩实验的流程和具体代码。此实验首先对模型所有卷积裁剪25%，同时进行蒸馏训练，然后进行离线量化。
```
Comment (qingqing01) [inline demo/auto-compression/pp-minilm/README.md]: 最开始出现PP-MiniLM加链接

File: demo/auto-compression/pp-minilm/run.py [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+    paddle.enable_static()
+
+    compress_config, train_config = load_config(args.config_path)
+    if train_config is not None and 'optim_args' in train_config:
+        train_config['optim_args'][
+            'apply_decay_param_fun'] = apply_decay_param_fun
+
+    train_dataloader, eval_dataloader = reader()
+    metric_class = METRIC_CLASSES[args.task_name]
+    metric = metric_class()
```
Comment (qingqing01) [inline demo/auto-compression/pp-minilm/run.py]: 这个metric在eval_function中采用到吧

File: demo/auto-compression/pp-minilm/run.py [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+from paddle.io import Dataset, BatchSampler, DataLoader
+from paddle.metric import Metric, Accuracy, Precision, Recall
+from paddlenlp.transformers import PPMiniLMForSequenceClassification, PPMiniLMTokenizer
+from paddlenlp.datasets import load_dataset
+from paddlenlp.data import Stack, Tuple, Pad
+from paddlenlp.data.sampler import SamplerHelper
+from paddlenlp.metrics import Mcc, PearsonAndSpearman
+from paddleslim.auto_compression.config_helpers import load_config
+from paddleslim.auto_compression.compressor import AutoCompression
+from utility import add_arguments, print_arguments
```
Comment (qingqing01) [inline demo/auto-compression/pp-minilm/run.py]: 清理上述没用的导入
这个文件夹下没有utility

**Review by yghstill — DISMISSED** [review]

File: demo/auto_compression/nlp/README.md [hunk 11]
```diff
@@ -0,0 +1,52 @@
+# 自然语言处理模型自动压缩示例
+
+本示例将介绍如何使用PaddleNLP中Inference部署模型进行自动压缩。
+
+## Benchmark
+- PP-MiniLM模型
```
Comment (wanghaoshuang) [inline demo/auto_compression/nlp/README.md:6]: 简要介绍下什么是PP-MiniLM, 并给个链接到详细介绍页。
至少让用户知道这是小型化的BERT结构。

File: demo/auto_compression/nlp/README.md [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+
+此实验首先对模型所有卷积裁剪25%，同时进行蒸馏训练，然后进行离线量化。
```
Comment (wanghaoshuang) [inline demo/auto_compression/nlp/README.md]: 是检测attention head吧？不是卷积。

Comment (wanghaoshuang) [inline demo/auto_compression/nlp/README.md]: "离线量化"后最好加个(Post-training quantization)。

File: demo/auto_compression/nlp/README.md [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+
+性能测试的环境为
+- 硬件：NVIDIA Tesla T4 单卡
+- 软件：CUDA 11.0, cuDNN 8.0, TensorRT 8.0
+- 测试配置：batch_size: 40, max_seq_len: 128
+
+## 环境准备
+
+### 1.准备数据
+本案例默认以CLUE数据进行自动压缩实验，如数据集为非CLUE格式数据，请修改启动文本run.sh中dataset字段。
```
Comment (wanghaoshuang) [inline demo/auto_compression/nlp/README.md]: 会自动下载CLUE数据么？可以明确告知用户。

File: demo/auto_compression/nlp/README.md [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+
+### 2.准备需要量化的环境
```
Comment (wanghaoshuang) [inline demo/auto_compression/nlp/README.md]: 建议直接给安装命令，并附上PaddlePaddle和PaddleNLP安装页面的链接.

File: demo/auto_compression/nlp/README.md [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+- PaddlePaddle >= 2.3
+- PaddleNLP >= 2.3
+
+注：安装PaddleNLP的目的是为了下载PaddleNLP中的数据集和Tokenizer。
+
+### 3.准备待压缩的部署模型
```
Comment (wanghaoshuang) [inline demo/auto_compression/nlp/README.md:51]: 请直接给出部署模型的下载链接

File: demo/auto_compression/nlp/README.md [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+如果已经准备好部署的model.pdmodel和model.pdiparams部署模型，跳过此步。
+根据[PaddleNLP文档](https://github.com/PaddlePaddle/PaddleNLP/tree/develop/examples)导出Inference模型，本示例可参考[PaddleNLP PP-MiniLM 中文小模型](https://github.com/PaddlePaddle/PaddleNLP/tree/develop/examples/model_compression/pp-minilm)微调后保存下每个数据集下有最高准确率的模型。
+
+## 开始自动压缩
+
+### 进行剪枝蒸馏和离线量化自动压缩
+
+蒸馏量化自动压缩示例通过run.py脚本启动，会使用接口paddleslim.auto_compression.AutoCompression对模型进行量化训练。将任务名称、模型类型、数据集名称、压缩参数传入，对模型进行剪枝、蒸馏训练和离线量化。数据集为CLUE，不同任务名称代表不同CLUE上的任务，可选择的任务名称有：afqmc, tnews, iflytek, ocnli, cmnli, cluewsc2020, csl。具体运行命令为：
```
Comment (wanghaoshuang) [inline demo/auto_compression/nlp/README.md]: “量化训练”？不是只有离线量化么？
“不同任务名称代表不同CLUE上的任务”->“不同任务名称代表CLUE上不同的任务”。

Comment (wanghaoshuang) [inline demo/auto_compression/nlp/README.md]: 行内代码建议用以下格式：
`paddleslim.auto_compression.AutoCompression`

File: demo/auto_compression/nlp/README.md [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+    --model_dir='./afqmc_base/' \
+    --model_filename='inference.pdmodel' \
+    --params_filename='inference.pdiparams' \
+    --dataset='clue' \
+    --save_dir='./save_afqmc_pruned/' \
+    --batch_size=16 \
+    --max_seq_length=128 \
+    --task_name='afqmc' \
+    --config_path='./configs/afqmc.yaml' 
+```
```
Comment (wanghaoshuang) [inline demo/auto_compression/nlp/README.md:152]: 1. 请简要概括介绍下`--config_path='./configs/afqmc.yaml' `的作用。
2. 另起一节或给出其它页面的链接，介绍如何修改剪枝或量化的参数配置。
3. 如果完全是自动的剪枝和量化，不支持用户自己设置相关参数，则显式说明，并重点介绍下afqmc.yaml中的训练参数。

File: demo/auto_compression/nlp/run.sh [hunk 18]
```diff
@@ -0,0 +1,16 @@
+export FLAGS_cudnn_deterministic=True
+export PYTHONPATH=/workspace/xuxu/auto_compress/add_demo/PaddleSlim/:$PYTHONPATH
+export PYTHONPATH=/workspace/xuxu/PaddleNLP/:$PYTHONPATH
+python3.7 run.py \
```
Comment (wanghaoshuang) [inline demo/auto_compression/nlp/run.sh]: 文档中写的`python`, 你这里用的`python3.7`, 建议对齐。

File: demo/auto_compression/nlp/run.sh [hunk 19]
```diff
@@ -0,0 +1,16 @@
+export FLAGS_cudnn_deterministic=True
+export PYTHONPATH=/workspace/xuxu/auto_compress/add_demo/PaddleSlim/:$PYTHONPATH
+export PYTHONPATH=/workspace/xuxu/PaddleNLP/:$PYTHONPATH
```
Comment (wanghaoshuang) [inline demo/auto_compression/nlp/run.sh]: 这两行都不需要吧，前提是已经安装好了PaddleSlim和PaddleNLP

File: demo/auto_compression/nlp/README.md [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+
+## 环境准备
+
+### 1.准备数据
+本案例默认以CLUE数据进行自动压缩实验，如数据集为非CLUE格式数据，请修改启动文本run.sh中dataset字段。
+
+### 2.准备需要量化的环境
+- PaddlePaddle >= 2.3
+- PaddleNLP >= 2.3
+
```
Comment (wanghaoshuang) [inline demo/auto_compression/nlp/README.md:30]: 也给一下PaddleSlim的安装方式和链接？

**Review by wanghaoshuang — CHANGES_REQUESTED** [review]
建议再完善下

File: demo/auto_compression/nlp/README.md [hunk 21]
```diff
@@ -0,0 +1,155 @@
...
+  activation_bits: 8
+  quantize_op_types:
+  - conv2d
+  - depthwise_conv2d
+  - mul
+  - matmul_v2
+  weight_bits: 8
+```
+
+### 进行剪枝蒸馏和离线量化自动压缩
```
Comment (wanghaoshuang) [inline demo/auto_compression/nlp/README.md:137]: 建议将该节提前到“压缩配置介绍”之前。

**Review by wanghaoshuang — APPROVED** [review]
